### PR TITLE
wait_for_dv_success: adapt to Populators, add stop_status_func optional arg

### PR DESCRIPTION
**Pull request by jpeimer**
##### Short description:

Changes in `wait_for_dv_success`:

1. Moving wait for PVC Bound to the end of the function
   because CSI storage is now using Populators for DataVolumes.
   With Populators, PVC gets Bound only after DV succeeded.

2. Adding `stop_status_func` optional arg (function) to exit if it returns True.

Changes in `DataVolume`:

1. Adding new DV phase `PendingPopulation`
   for DVs provisioned on CSI WFFC storage.

2. Adding timeout constants

### CODE REVIEW

File: ocp_resources/datavolume.py [hunk 1]
```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
                     continue
                 break
         except TimeoutExpiredError:
-            self.logger.error(f"{self.name} status is {sample}")
+            self.logger.error(f"{self.name} status is {sample.status}")
```
Comment (myakove) [inline ocp_resources/datavolume.py:277]: `sample.status` will fail if `sample` is `None`

File: ocp_resources/datavolume.py [hunk 2]
```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
...
             raise
 
-    def wait_for_dv_success(self, timeout=600, failure_timeout=120):
+    def _check_dv_progress(self):
+        dv_status = self.instance.status
+        dv_status_conditions = dv_status.get("conditions")
```
Comment (myakove) [inline ocp_resources/datavolume.py:282]: `dv_status.get("conditions")` > `dv_status.conditions`

File: ocp_resources/datavolume.py [hunk 3]
```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
...
+        if (
+            dv_status.restartCount
+            and dv_status.restartCount > 3
+            and dv_status_conditions
+            and dv_status_conditions[-1].type == "Running"
```
Comment (myakove) [inline ocp_resources/datavolume.py:287]: Why do you rely on the condition order `[-1]` ?

File: ocp_resources/datavolume.py [hunk 4]
```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
...
         except TimeoutExpiredError:
-            self.logger.error(f"Status of {self.kind} {self.name} is {sample}")
+            self.logger.error(
+                f"Status of {self.kind} '{self.name}' in '{self.namespace}' namespace is \n"
+                f"{sample.status}"
+            )
             raise
 
+        # For CSI storage, PVC gets Bound after DV succeeded
+        return self.pvc.wait_for_status(
```
Comment (myakove) [inline ocp_resources/datavolume.py:340]: this return don't belong to this function which only wait for dv (`wait_for_dv_success`)

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/datavolume.py (hunk 4 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py:340]: we can't say DV succeeded if PVC is not Bound.
And if DV was garbage collected - Bounded PVC is our only evidence.

File: ocp_resources/datavolume.py (hunk 3 above)
Comment (myakove) [reply ocp_resources/datavolume.py:287]: we don't needto rely on order, please change it

File: ocp_resources/datavolume.py [hunk 5]
```diff
@@ -269,29 +274,73 @@ def _check_none_pending_status(self, failure_timeout=120):
...
                 wait_timeout=timeout,
                 func=lambda: self.exists,
             ):
-                # DV reach to success if the status is succeeded or if the DV does not exists
+                # DV reach to success if the status is succeeded or if the DV does not exist
                 if sample is None or sample.status.phase == self.Status.SUCCEEDED:
-                    return
+                    break
+                elif check_progress:
+                    self._check_dv_progress()
```
[the remark this answers is not in the document]
Comment (myakove) [reply ocp_resources/datavolume.py:330]: I'm not fully understand what are you trying to do, I think it's better to talk :)

File: ocp_resources/datavolume.py (hunk 3 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py]: changing it, and moving this function out of here

File: ocp_resources/datavolume.py [hunk 6]
```diff
@@ -272,26 +277,62 @@ def _check_none_pending_status(self, failure_timeout=120):
             self.logger.error(f"{self.name} status is {sample}")
             raise
 
-    def wait_for_dv_success(self, timeout=600, failure_timeout=120):
+    def wait_for_dv_success(
+        self,
+        timeout=TIMEOUT_10MINUTES,
+        failure_timeout=TIMEOUT_2MINUTES,
+        stop_status=None,
```
Comment (myakove) [inline ocp_resources/datavolume.py:284]: `stop_status` is a function, we need to get *args and **kwargs for it.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/datavolume.py [hunk 7]
```diff
@@ -272,26 +277,62 @@ def _check_none_pending_status(self, failure_timeout=120):
...
-                # DV reach to success if the status is succeeded or if the DV does not exists
+                # DV reach to success if the status is succeeded or if the DV does not exist
                 if sample is None or sample.status.phase == self.Status.SUCCEEDED:
-                    return
+                    break
+                elif stop_status and stop_status(self):
+                    raise TimeoutExpiredError(
+                        value=(
+                            f"Exited on the stop_status True"
+                            f"{status_of_dv} is \n{sample.status}"
```
Comment (rnetser) [inline ocp_resources/datavolume.py:323]:
```suggestion
                            f"{status_of_dv} is {sample.status}"
```

File: ocp_resources/datavolume.py [hunk 8]
```diff
@@ -272,26 +277,62 @@ def _check_none_pending_status(self, failure_timeout=120):
...
+                        )
+                    )
+
         except TimeoutExpiredError:
-            self.logger.error(f"Status of {self.kind} {self.name} is {sample}")
+            self.logger.error(f"{status_of_dv} is \n{sample.status}")
```
Comment (rnetser) [inline ocp_resources/datavolume.py:328]: ditto

File: ocp_resources/datavolume.py [hunk 9]
```diff
@@ -272,26 +277,62 @@ def _check_none_pending_status(self, failure_timeout=120):
...
+        Returns:
+            bool: True if DataVolume succeeded.
+        """
         self.logger.info(f"Wait DV success for {timeout} seconds")
         self._check_none_pending_status(failure_timeout=failure_timeout)
-        self.pvc.wait_for_status(
-            status=PersistentVolumeClaim.Status.BOUND, timeout=timeout
+
+        sample = None
+        status_of_dv = (
```
Comment (rnetser) [inline ocp_resources/datavolume.py:307]:
```suggestion
        status_of_dv_str = (
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/datavolume.py [hunk 10]
```diff
@@ -272,26 +277,64 @@ def _check_none_pending_status(self, failure_timeout=120):
             self.logger.error(f"{self.name} status is {sample}")
             raise
 
-    def wait_for_dv_success(self, timeout=600, failure_timeout=120):
+    def wait_for_dv_success(
+        self,
+        timeout=TIMEOUT_10MINUTES,
+        failure_timeout=TIMEOUT_2MINUTES,
+        stop_status=None,
+        *args,
```
Comment (myakove) [inline ocp_resources/datavolume.py:285]:
```*args,
        **kwargs,
```
please rename to `*stop_status_args` and `**stop_status_kwargs`

and use them (not used right now)

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/datavolume.py (hunk 10 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py:285]: I'm not quite sure we need them at all. I expect that the passed function will do all the checks, and we don't know how this function will look like and what arguments should it get.

**Comment by myakove** [comment]
Please update the PR title and the `Short description` to match your changes

File: ocp_resources/datavolume.py [hunk 11]
```diff
@@ -272,26 +277,64 @@ def _check_none_pending_status(self, failure_timeout=120):
...
+        **kwargs,
+    ):
+        """
+        Wait until DataVolume succeeded with or without DV Garbage Collection enabled
+
+        Args:
+            timeout (int):  Time to wait for the DataVolume to succeed.
+            failure_timeout (int): Time to wait for the DataVolume to have not Pending/None status
+            stop_status (function): function that is called inside the TimeoutSampler
```
Comment (oharan2) [inline ocp_resources/datavolume.py:294]: Could we please rename `stop_status` to be more intuitive as it's a function?
Examples:

stop/break`_sample_function`
sample/check`_block_function`
`stop_status_func` etc...

File: ocp_resources/datavolume.py (hunk 10 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py]: Done, thanks @myakove, @rnetser for your help!

File: ocp_resources/datavolume.py (hunk 11 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py]: Done

File: ocp_resources/datavolume.py [hunk 12]
```diff
@@ -272,26 +277,69 @@ def _check_none_pending_status(self, failure_timeout=120):
             self.logger.error(f"{self.name} status is {sample}")
             raise
 
-    def wait_for_dv_success(self, timeout=600, failure_timeout=120):
+    def wait_for_dv_success(
+        self,
+        timeout=TIMEOUT_10MINUTES,
+        failure_timeout=TIMEOUT_2MINUTES,
+        stop_status_func=None,
+        *stop_status_args,
```
Comment (myakove) [inline ocp_resources/datavolume.py:285]: *stop_status_func_args, also for kwargs.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/datavolume.py (hunk 12 above)
Comment (jpeimer) [reply ocp_resources/datavolume.py]: Done

**Review by rnetser — APPROVED** [review]

**Pull request merged**